### PR TITLE
Proof and correctness logic reorganization

### PR DESCRIPTION
**Pull request by jMyles**
Bob needs a way to verify the correctness of individual cfrags.  This gives him that.

Also, I undertook a substantial rewrite of both `verify_correctness` and `_prove_correctness` in an effort to bring them closer to together, so that the reader may more easily understand how they inter-operate.

### CODE REVIEW

File: umbral/pre.py [hunk 1]
```diff
@@ -286,39 +287,44 @@ def reencrypt(kfrag: KFrag, capsule: Capsule, params: UmbralParameters=None,
 def _prove_correctness(cfrag: CapsuleFrag, kfrag: KFrag, capsule: Capsule, 
                        metadata: bytes=None, params: UmbralParameters=None
                       ) -> CorrectnessProof:
-    params = params if params is not None else default_params()
+    t = CurveBN.gen_rand(params.curve)
+    rk = kfrag._bn_key
 
-    e1 = cfrag._point_e1
-    v1 = cfrag._point_v1
+    params = params if params is not None else default_params()
```
Comment (cygnusv) [inline umbral/pre.py]: `params` initialization should be done in the first line. I would also prefer to keep `t = CurveBN.gen_rand(params.curve)` just in the previous line to its first use `e2 = t * e`, but that's a matter of personal preference.

File: umbral/fragments.py [hunk 2]
```diff
@@ -225,7 +227,61 @@ def to_bytes(self):
...
+    def verify_correctness(self, capsule: "Capsule",
+                       pubkey_a: UmbralPublicKey, pubkey_b: UmbralPublicKey,
+                       params: UmbralParameters=None):
+
+        params = params if params is not None else default_params()
+        pubkey_a_point = pubkey_a.point_key
+        pubkey_b_point = pubkey_b.point_key
+
+        ####
+        ## Here are the formulaic constituents shared with `prove_correctness`.
```
Comment (cygnusv) [inline umbral/fragments.py]: "... with `pre._prove_correctness`"

File: umbral/fragments.py [hunk 3]
```diff
@@ -225,7 +227,61 @@ def to_bytes(self):
 
         return serialized_cfrag
 
-    def __bytes__(self):
-        return self.to_bytes()
+    def verify_correctness(self, capsule: "Capsule",
+                       pubkey_a: UmbralPublicKey, pubkey_b: UmbralPublicKey,
+                       params: UmbralParameters=None):
+
```
Comment (cygnusv) [inline umbral/fragments.py]: I must say that I prefer to keep the logic in `umbral.pre` and to have this only as a wrapper (didn't we agree on Discord to wrap this?). Now that I think about it, it's the same for `KFrag.verify`.

File: umbral/pre.py [hunk 4]
```diff
@@ -252,7 +252,8 @@ def split_rekey(priv_a: Union[UmbralPrivateKey, CurveBN],
         u1 = rk * u
         y = CurveBN.gen_rand(params.curve)
 
-        z1 = CurveBN.hash(y * g, id_kfrag, pub_a, pub_b, u1, xcomp, params=params)
+        signature_input = [y * g, id_kfrag, pubkey_a_point, pubkey_b_point, u1, xcomp]
+        z1 = CurveBN.hash(*signature_input, params=params)
         z2 = y - priv_a * z1
```
Comment (cygnusv) [inline umbral/pre.py]: Same TODO above this block

File: umbral/fragments.py (hunk 3 above)
Comment (tuxxy) [reply umbral/fragments.py]: Yeah, I think I prefer this too. Just to keep the logic where the rest of the cryptographic code is.

@cygnusv Am I right that all of the logic described in the Umbral paper is in `umbral.pre`?

Comment (jMyles) [reply umbral/fragments.py]: Ahh, right, we did indeed agree to wrap this - that slipped my mind.

As for keeping all the "cryptographic" code together: it's far from obvious how to do that.  There's code reasonable understood as "cryptography" in every single module, including this one.  

But as for wrapping this, I'll mark this [WIP] and get a fix in later today.  :-)

File: umbral/fragments.py (hunk 2 above)
Comment (jMyles) [reply umbral/fragments.py]: :+1:

File: umbral/pre.py (hunk 1 above)
Comment (jMyles) [reply umbral/pre.py]: Right - `params` was a goof here.  

On `t`, I had the same inclination, but then that makes it impossible to do a line-for-line match with the verify logic.

File: umbral/fragments.py (hunk 3 above)
Comment (cygnusv) [reply umbral/fragments.py]: Well, I was talking about Umbral KEM itself. Of course there's cryptographic code in `umbral.point` or `umbral.openssl`, but that's a different abstraction layer, in particular, a lower level. If we move the logic of `CFrag.verify_correctness` and `KFrag.verify` to `umbral.pre`, then I think all the core cryptographic logic of Umbral is in `umbral.pre`, in approximately 600 lines of code. That's neat, IMHO.

**Comment by cygnusv** [comment]
@tuxxy Not all, but most of it. What parts of Umbral are not in `umbral.pre`?
- Key generation is in `umbral.keys`, but that's reasonable since Umbral encryption keys follow the same simple approach than most schemes based on ElGamal/Diffie-Hellman (i.e., `pk = sk * g).
- Some auxiliary stuff in `umbral.util`, which is basically non-cryptographic code.

**Pull request merged**